crates/red_knot_python_semantic/src/types.rs

```
@@ -1953,8 +1953,11 @@ impl<'db> Type<'db> {
                         let (ty_a, ty_b) = binding
                             .two_parameter_tys()
                             .unwrap_or((Type::unknown(), Type::unknown()));
-                        binding
-                            .set_return_ty(Type::BooleanLiteral(ty_a.is_equivalent_to(db, ty_b)));
+
+                        let equivalence =
+                            ty_a.is_equivalent_to(db, ty_b) && ty_b.is_equivalent_to(db, ty_a);
```

I don't think this change is a good idea.

In the previous tests, we asserted the relation in both directions for every pair of types, including in the negative cases. This change doesn't give us similar testing of commutativity, particularly for the negative tests, because if one direction returns true and the other direction returns false, we'll just report false. So non-commutativity in the negative equivalence tests would have been caught as an error previously, and now would silently pass.

I also think that for clarity reasons, `knot_extensions.is_equivalent_to` should simply reflect what our `is_equivalent_to` function does, and not implicitly double-check the commutation.

I think that instead of making this change, we should explicitly test both directions in the markdown tests. If that's too painful, we should consider just not moving these to mdtests.

@sharkdp, you already reviewed this and were prepared to land it as-is; how do you feel about this?

---

Good point about the `is_disjoint_from` tests!

I feel that `knot_extensions.is_equivalent_to` and `knot_extensions.is_disjoint_from` are semantically meaningful type operations that I'm comfortable exposing publicly to users of red-knot. These operations should simply be commutative, and if they are not, it's a bug. Testing that we in fact implement them commutatively is a concern only of red-knot's test suite, which I don't want to expose in the public API of those methods.

So if we want to move these tests to mdtest, without manual testing of commutativity, then I prefer Alex's suggestion of dedicated methods for this purpose.

My main hesitance is that these would be the first methods in the `knot_extensions` namespace that are intended solely for testing red-knot, and that I would not prefer to expose otherwise. This suggests possibly putting them in a different namespace, and possibly adding some infrastructure to only expose that namespace in red-knot tests, and not otherwise. I think this is all doable, but it's certainly more work.

Would like to hear @sharkdp take on this.

---

Oh! Last time I looked at the PR, it was still showing a version that was combined with the is-subtype-of changes, and I completely missed this change, thinking it was something from the previous changeset.


> I also think that for clarity reasons, `knot_extensions.is_equivalent_to` should simply reflect what our `is_equivalent_to` function does,

Yes, absolutely.

> Perhaps `is_equivalent_to`/`is_disjoint_from` should return `tuple[bool, bool]` instead?

> or we could add more functions such as `knot_extensions.is_commutatively_equivalent()` and `knot_extensions.is_commutatively_disjoint()`?

I'm personally not a big fan of either of these options. I would propose we add a "is_equivalent_to is commutative" section to this suite, add a couple of (non exhaustive) tests, and then rely on a `is_equivalent_to_is_commutative` property test to specifically test this property. We can add a reference to the property test in this file.

Yes, this would not be equivalent to (heh) what we had before, but I don't think that a commutativity-violation is a very likely scenario. And if so, we would hopefully catch that using the new property test.

---

That reminds me that I had an idea for a way we could issue a warning to users about the `knot_extensions` API relatively easily: https://github.com/astral-sh/ruff/issues/15355#issuecomment-2590811421

---

I'm also happy to go with @sharkdp's suggested path here!

---

I like @sharkdp 's proposal.

---

Additional thought: if we had "inlining" of small functions during type inference (like pyright), we could potentially even write helpers such as `is_commutatively_equivalent` in Python, which would be neat.

---

Either inlining (so we could just write the helpers without annotations), or explicit `TypeForm` annotations, would allow this!

---

https://github.com/astral-sh/ruff/pull/15488

---

> or explicit `TypeForm` annotations

For posterity: @sharkdp pointed out in chat that this wouldn't be enough to propagate the right literal boolean return type back; we'd need to be able to annotate with something like `def is_commutative_equivalent_to[T, S](ty_a: TypeForm[T], ty_b: TypeForm[S]) -> And[is_equivalent_to(T, S), is_equivalent_to(S, T)]: ...` -- obviously that return type annotation is depending on several things that don't exist :)